README
NAME
    IO::Event - Tied Filehandles for Nonblocking IO with Object Callbacks

SYNOPSIS
     use IO::Event;
     use IO::Event 'emulate_Event';

     my $ioe = IO::Event->new($filehandle);

     my $ioe = IO::Event::Socket::INET->new( [ARGS] )

     my $ioe = IO::Event::Socket::UNIX->new( [ARGS] )

     my $timer = IO::Event->timer(interval => $seconds);

     IO::Event::loop();

     IO::Event::unloop_all();

DESCRIPTION
    IO::Event provides a object-based callback system for handling
    nonblocking IO. The design goal is to provide a system that just does
    the right thing w/o the user needing to think about it much.

    All APIs are kept as simple as possible yet at the same time, all
    functionality is accesible if needed. Simple things are easy. Hard
    things are possible.

    Most of the time file handling syntax will work fine: "<$filehandle>"
    and "print $filehandle 'stuff'".

    After initial setup, call "IO::Event::loop()".

    IO::Event was originally written to use Event. It turns out that Event
    has bugs and now IO::Event can provide it's own select loop instead of
    using the one from Event.

CONSTRUCTORS
    IO::Event->new($filehandle, [ $handler, [ $options ]])
        The basic "new" constructor takes a filehandle and returns a
        psuedo-filehandle. Treat the IO::Event object as a filehandle. Do
        not use the original filehandle without good reason (let us know if
        you find a good reason so we can fix the problem).

        The handler is the class or object where you provide callback
        functions to handle IO events. It defaults to the package of the
        calling context.

        If present, $options is a hash reference with the following possible
        keys:

        description  A text description of this filehandle. Used for
                     debugging and error messages.

        read_only    Set to true if this is a read-only filehandle. Do not
                     accept output.

        write_only   Set to true if this is a write-only filehandle. Do not
                     attept to read.

        autoread     Set to 0 if this should not be an auto-read filehandle.

    IO::Event::Socket::INET->new( [ARGS] )
        This constructor uses IO::Socket::INET->new() to create a socket
        using the ARGS provided. It returns an IO::Event object.

        The handler defaults as above or can be set with an additional
        pseudo-parameter for IO::Socket::UNIX->new(): "Handler". A
        description for the socket can be provided with an additional
        psuedo-parameter: "Description".

    IO::Event::Socket::UNIX->new( [ARGS] )
        This constructor uses IO::Socket::UNIX->new() to create a socket
        using the ARGS provided. It returns an IO::Event object.

        The handler defaults as above or can be set with an additional
        pseudo-parameter for IO::Socket::UNIX->new(): "Handler". A
        description for the socket can be provided with an additional
        psuedo-parameter: "Description".

MANDATORY HANDLERS
    These handler methods must be available in the handler object/class if
    the situation in which they would be called arises.

    ie_input($handler, $ioe, $input_buffer_reference)
        Invoked when there is fresh data in the input buffer. The input can
        be retreived via directly reading it from $$input_buffer_reference
        or via "read()" from the $ioe filehandle, or by using a variety of
        standard methods for getting data:

                <$ioe>                  like IO::Handle
                $ioe->get()             like Data::LineBuffer
                $ioe->read()            like IO::Handle
                $ioe->sysread()         like IO::Handle
                $ioe->getline()         like IO::Handle
                $ioe->getlines()        like IO::Handle
                $ioe->getsome()         see below
                $ioe->ungets()          like FileHandle::Unget

        At end-of-file, ie_input will only be invoked once. There may or may
        not be data in the input buffer.

    ie_connection($handler, $ioe)
        Invoked when a listen()ing socket is ready to accept(). It should
        call accept:

                sub ie_connection
                {
                        my ($pkg, $ioe) = @_;
                        my $newfh = $ioe->accept()
                }

    ie_read_ready($handler, $ioe, $underlying_file_handle)
        If autoreading is turned off then this will be invoked.

    ie_werror($handler, $ioe, $output_buffer_reference)
        A write error has occured when trying to drain the write buffer.
        Provide an empty subroutine if you don't care.

OPTIONAL HANDLERS
    These handler methods will be called if they are defined but it is not
    required that they be defined.

    ie_eof($handler, $ioe, $input_buffer_reference)
        This is invoked when the read-side of the filehandle has been closed
        by its source.

    ie_output
        This is invoked when data has just been written to the underlying
        filehandle.

    ie_outputdone
        This is invoked when all pending data has just been written to the
        underlying filehandle.

    ie_connected
        This is invoked when a "connect()" completes.

    ie_connect_failed($handler, $ioe, $error_code)
        This is invoked when a "connect()" fails. For a timeout, the error
        code will be ETIMEOUT.

    ie_died($handler, $ioe, $method, $@)
        If another handler calls "die" then ie_died will be called with the
        IO::Event object, the name of the method just invoked, and the die
        string. If no ie_died() callback exists then execution will
        terminate.

    ie_timer
        This is invoked for timer events. These will only arise if set using
        thing underlying Event object.

    ie_exception
        Invoked when an exceptional condition arises on the underlying
        filehandle

    ie_outputoverflow($handler, $ioe, $overflowing,
    $output_buffer_reference)
        Invoked when there is too much output data and the output buffers
        are overflowing. You can take some action to generate less output.
        This will be invoked exactly once (with $overflowing == 1) when
        there is too much data in the buffer and then exactly once again
        (with $overflowing == 0) when there is no longer too much data in
        the buffer.

METHODS
    In addition to methods described in detail below, the following methods
    behave like their "IO" (mostly "IO::Socket") counterparts (except for
    being mostly non-blocking...):

            connect
            listen
            open
            read
            sysread
            syswrite
            print
            eof

    Through AUTOLOAD (see the SUBSTITUTED METHODS section) methods are
    passed to underlying "Event" objects:

            loop
            unloop
            and many more...

    Through AUTOLOAD (see the SUBSTITUTED METHODS section) methods are
    passed to underlying "IO" objects:

            fileno
            stat
            truncate
            error
            opened
            untaint
            and many more...

    IO::Event defines its own methods too:

    ->accept($handler, %options)
        accept() is nearly identical to the normal IO::Socket::accept()
        method except that instead of optionally passing a class specifier
        for the new socket, you optionally pass a handler object or class.
        The returned filehandle is an IO::Event object.

        Supported options:

        description
            Sets the description for the new socket

        autoread
            Set to 0 if you do not want auto-read

    ->can_read($amount)
        Returns true if $amount bytes worth of input is available for
        reading. Note: this does not return true at EOF so be careful not to
        hang forever at EOF.

    ->getsome($amount)
        Returns $amount bytes worth of input or undef if the request can't
        be filled. Returns what it can at EOF.

    ->get()
        get() is like getline() except that it pre-chomp()s the results and
        assumes the input_record_separator is "\n". This is like get() from
        Data::LineBuffer.

    ->unget()
        Push chomp()ed lines back into the input buffer. This is like
        unget() from Data::LineBuffer.

    ->ungetline(), ->xungetc(), ->ungets()
        This is what ungetc() should be: it pushes a string back into the
        input buffer. This is unlike IO::Handle->ungetc which takes an
        ordinal and pushes one character back into the the input buffer.

    ->handler($new_handler)
        Sets the handler object/class if $new_handler is provided. Returns
        the old handler.

    ->filehandle()
        Returns the underlying "IO::Handle".

    ->event()
        Returns the underling "Event".

    ->listener($listening)
        Used to note that a filehandle is being used to listen for
        connections (instead of receiving data). A passed parameter of 0
        does the opposite. Returns the old value. This is mostly used
        internally in IO::Event.

    ->input_record_separator($new_sep)
        IO::Handle doesn't allow input_record_separator's on a per
        filehandle basis. IO::Event does. If you don't ever set a
        filehandle's input record separator, then it contineously defaults
        to the current value of $/. If you set it, then it will use your
        value and never look at $/ again.

    ->readevents($readevents)
        Get/set listening for read-ready events on the underlying
        filehandle. This could be used by ie_outputoverflow to control input
        flows.

    ->output_bufsize($output_bufsize)
        Get/set the size of the output buffer.

    ->autoread($autoread)
        Get/set automatic reading if data when data can be read. Without
        autoread turned on, the input buffer ins't filled and none of the
        read methods will work. The point of this is for working with
        non-data filehandles. This is an experts-only method that kinda
        defeats the purpose of this module. This would be necessary using
        recv() to get data.

    ->drain()
        Used to start looking for write-ready events on the underlying
        filehandle. In normal operation this is handled automatically.
        Deprecated: use writeevents(1) instead.

    ->reentrant($reentrant)
        Get/set reentrant callbacks. By default, IO::Event avoids making
        reentrant callbacks. This is good because your code is less likely
        to break. This is bad because you won't learn about things right
        away. For example, you will not learn the the output buffer is
        overflowing during print(). You'll have to wait for the output
        buffer to begin draining to find out. This could be a problem.

    ->close()
        If there is output buffered, close will be delayed until the output
        buffer drains.

    ->forceclose
        Close close immediately, even if there is output buffered.

SUBSTITUED METHODS
    Any method invocations that fail because the method isn't defined in
    IO::Event will by tried twice more: once using trying for a method on
    the inner (hidden) filehandle and once more trying for a method on the
    Event object that's used to create the select loop for this module.

EXAMPLE SERVER
            # This is a tcp line echo server

            my $listener = IO::Event::Socket::INET->new(
                    Listen => 10,
                    Proto => 'tcp',
                    LocalPort => 2821,
            );

            Event::loop();

            sub ie_connection
            {
                    my ($pkg, $lstnr) = @_;
                    my $client = $lstnr->accept();
                    printf "accepted connection from %s:%s\n",
                            $client->peerhost, $client->peerport;
            }

            sub ie_input
            {
                    my ($pkg, $client, $ibufref) = @_;
                    print $client <$client>;
            }

SYSREAD and EOF
    sysread() is incompatable with eof() because eof() uses getc(). Most of
    the time this isn't a problem. In other words, some of the time this is
    a problem: lines go missing.

    For this reason, IO::Event never uses sysread(). In fact, if you ask it
    to do a sysread() it does a read() for you instead.

    On the other hand, at the current time no problems with syswrite have
    come to light and IO::Event uses syswrite and never any other form of
    write/print etc.

AVOIDING Event.pm
    If you find that your IO::Event program works fine for a while and then
    slows down or fails to receive all the events you expect, the problem
    may be in Event. IO::Event can optionally use it's own select loop
    instead of Event.

    To use IO::Event's select loop, import "emulate_Event".

     use IO::Event 'emulate_Event';

    IO::Event does not provide a complete emulation of everything that Event
    does. It provides the full timer API:

     my $timer = IO::Event::timer( [ARGS] )

    instead of

     my $timer = Event::timer( [ARGS] )

    However it does not provide timer events on filehandles, nor does it
    provide events for signals, or variable accesses.

    Use "IO::Event::loop()" instead of "Event::loop()". Use
    "IO::Event::unloop_all()" instead of "Event::unloop_all()". It does not
    provide any other methods or functions from Event. If you need them,
    please send a patch.

    IO::Event's definition for "loop()", "timer()", and "unloop_all()"
    defaults to the Event version unless "emulate_Event" has been imported.
    This allows you to easily switch back and forth between Event's
    dispatcher and IO::Event's version. IO::Event's version is generally
    lower performance but it seems to be more reliable.

DATA STRUCTURE
    The filehandle object itself is a funny kind of hash reference. If you
    want to use it to store your own data, you can. Please don't use hash
    keys that begin "ie_" or "io_" as those are the prefixes used by
    "IO::Event" and "IO::Socket".

    The syntax is kinda funny:

            ${*$filehandle}{'your_hash_key'}  

SEE ALSO
    The following perl modules do something that is kinda similar to what is
    being done here:

            IO::Multiplex
            IO::NonBlocking
            IO::Select
            Event
            POE
            POE::Component::Server::TCP
            Net::Socket::NonBlock
            Net::Server::Multiplex
            NetServer::Generic

    The API borrows most heavily from IO::Multiplex. IO::Event uses Event.pm
    and thus can be used in programs that are already using Event or POE.

BUGS
    The test suite only covers 40% of the code.

LICENSE
    Copyright (C) 2002-2009 David Muir Sharnoff <muir@idiom.org>. This
    module may be used/copied/etc on the same terms as Perl itself.



CHANGELOG

Added missing depenency on List::MoreUtils

= 2013/04/04

Remove rinetd.pl -- becoming its own module

= 2013/03/20

No longer required to start with IO::Event->new()

fixed forked3.t, getline3.t, and multifork3.t tests

Added timer tests.

Now specifies "hard" on Event.pm timers

Added ->ie_desc()

= 2013/03/14	0.807

Bugfix: IO::Event constructor calls that did not include an
explicit handler were not getting the correct handler.

= 2012/02/28	0.806

Add 'diagnostics' to build requirements.

Documentation updates.

= 2011/06/17	0.805

Docfix.  It's IO::Event->timer() not IO::Event::timer.

= 2011/02/28	0.804

Adjust tests to use 127.0.0.1 instead of Sys::Hostname::hostname 

= 2011/02/27	0.803

Adjust prerequisites and tests.  Require Event & AnyEvent, run tests w/o them anyway.

= 2011/02/26	0.802

Add missing files to the MANIFEST
Forgot to update README

= 2011/02/25	0.801

Add framework to support multiple backend event loops.
Add AnyEvent as a new supported event loop.

= 2009/07/23	0.704

Added new API: IO::Event::Callback that provides a 
closure/callback interface instead of the subclass
interface of IO::Event.  It's a layer.

= 2009/04/29	0.703

OOps.  Changes for 0.702 didn't actually make it in.

= 2009/04/28	0.702

Added shutdown()

Bugfix: writes when the OS buffer was completely full are now handled properly

= 2009/02/17	0.701

Made the use of Event.pm optional

= 2009/01/30	0.603

Write buffering debugged.

Failure to close() on eof() now auto-closes.

Rewrote rinetd.pl to use Daemon::Generic

= 2006/10/05	0.602

Minor bugfix to rinetd.pl

= 2005/11/04	0.601

Added example program: rinetd.pl

API CHANGE: 
Replaced ie_connect_timeout() with ie_connect_failed().  
Previously ie_eof() would be called for non-timeout connection
failures.  Now ie_connect_failed() handles all connection failues.

Documentation fixes.

= 2005/08/29	0.508

Bugfix: don't try to read recently-closed filehandles.

= 2005/08/29	0.507

Add support for read-only and write-only descriptors.

= 2005/01/27	0.506

Drop Event objects to prevent memory leaks.

= 2005/01/21	0.505

Removed some race-condition warnings if descriptors are
close()ed before EOF.

= 2004/05/05	0.504

Dropped all calls to eof().  Switched from read() to
sysread().  Fixed bugs in ie_invoke.  Made forked.t
portable to DragonflyBSD which probably also made it
work on a bunch of other systems too.

= 2004/04/20	0.503

Bugfix: eof was falsing positive

Bugfix: partial lines were being returned for <$ieo>

= 2004/02/24	0.502

Added can_read()

Bugfix: spurrious warning in unget()

Added ungets() to support FileHandle::Unget semantics

= 2003/11/29	0.501

Changed version numbering 'cause we don't want to get to 1.0 quite yet.

Add reentrantcy controls 

Add readevents()

Add output buffer size related code.

Bugfix: remove input watermark code
Bugfix: retry read() until EAGAIN prevents lost events
Bugfix: don't suppress ie_eof after reopen or connect.

= 2003/11/26	0.5

Lots more bugs fixed.  Documentation added.  Test case added.

= 2003/11/26	0.4

Lots of bad bugs fixed.  Some speedups.

= 2003/11/26	0.3

Fixed the semantics of read() and sysread() to match perl's 
documentation.  Oops.

Added 'use diagnostics' and fixed the things it complained about.

Added descriptions to the test suite.

= 2003/04/14	0.2

Added sysread()

= 2002/12/20	0.1

Initial revision


Makefile.PL

use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile being created.
WriteMakefile(
	'NAME'		=> 'IO::Event',
	'DISTNAME'	=> 'IO-Event',
	'VERSION_FROM'	=> 'Event.pm',
	'EXE_FILES'	=> ['rinetd.pl'],
	($] >= 5.005 ?
	    ('ABSTRACT' => 'Tied Filehandles for Nonblocking IO with Object Callbacks',
	     'AUTHOR'	=> 'David Muir Sharnoff <muir@idiom.com>') : ()),
	'dist'		=> {COMPRESS=>'gzip', SUFFIX=>'gz'},
	'PREREQ_PM'	=> {
		'Event' => 0.87,
		'IO::Handle' => 0,
		'Time::HiRes' => 0,
		'Test::Simple' => 0,

		'Getopt::Long' => 0,
		'File::Slurp' => 0,
		'File::Flock' => 0,
		'Net::Netmask' => 0,
		'Daemon::Generic' => 0,
	 },
);

package MY;

sub postamble {
    <<"END_OF_POSTAMBLE";

pm_to_blib: README

README: Event.pod
\tpod2text Event.pod >README

END_OF_POSTAMBLE
}


rinetd.pl
#!/usr/bin/perl

my $pidbase = '/var/run/rinetd';
my $configfile = '/etc/rinetd.pl.conf';
my $logger_args = "-t $0";
my $debug = 0;

$main::VERSION = '1.1';

use strict;
use IO::Event;
use Net::Netmask;
use Getopt::Long;
use File::Slurp;
use File::Flock;
require POSIX;

my %config;
my %filters;
my $counter = 1;
sub error;

my $foreground = 0;

Getopt::Long::Configure("auto_version");
GetOptions(
	'configfile=s'	=> \$configfile,
	'foreground!'	=> \$foreground,
) or usage();

usage() if @ARGV != 1;

my $do = shift(@ARGV);

if ($do eq 'help') {
	usage();
} elsif ($do eq 'version') {
	print "$0 - version $main::VERSION\n";
	exit;
} 

my $pidfile;
{
	my $x = $configfile;
	$x =~ s!/!.!g;
	$pidfile = "$pidbase$x.pid";
}

my %newconfig = preconfig();

print "Configuration looks okay\n" if $do eq 'check';

my $killed = 0;
my $locked = 0;
if (-e $pidfile) {
	if ($locked = lock($pidfile, undef, 'nonblocking')) {
		# old process is dead
	} else {
		sleep(2) if -M $pidfile < 2/86400;
		my $oldpid = read_file($pidfile);
		chomp($oldpid);
		if ($oldpid) {
			if ($do eq 'stop' or $do eq 'restart') {
				kill(2,$oldpid) and print "Killed $oldpid\n";
				exit if $do eq 'stop';
				$locked = lock($pidfile);
				$killed = 1;
			} elsif ($do eq 'reload') {
				if (kill(1,$oldpid)) {
					print "Requested reconfiguration\n";
					exit;
				} else {
					print "Kill failed: $!\n";
				}
			} elsif ($do eq 'check') {
				if (kill(0,$oldpid)) {
					print "$0 running - pid $oldpid\n";
					exit;
				} 
			} else {
				error "Exiting $0 processing running, use stop, restart, or reload\n";
			}
		} else {
			error "Pid file $pidfile is invalid but locked, exiting\n";
		}
	}
} else {
	$locked = lock($pidfile, undef, 'nonblocking') 
		or die "Could not lock pid file $pidfile: $!";
}

if ($do eq 'reload' || $do eq 'stop' || $do eq 'check' || ($do eq 'restart' && ! $killed)) {
	print "No $0 running\n";
}

exit if $do eq 'stop';
exit if $do eq 'check';


usage() unless $do eq 'reload' || $do eq 'restart' || $do eq 'start';

unless ($foreground) {
	print "Starting $0 server\n";
	my $pid;
	POSIX::_exit(0) if $pid = fork;
	die "Could not fork: $!" unless defined $pid;
	POSIX::_exit(0) if $pid = fork;
	die "Could not fork: $!" unless defined $pid;

	POSIX::setsid();

	open(STDERR, "|logger $logger_args");
} else {
	print "Starting...\n";
}

$locked or lock($pidfile, undef, 'nonblocking') 
	or die "Could not lock pid file $pidfile: $!";

write_file($pidfile, "$$\n");

select(STDERR);
$| = 1;

postconfig(%newconfig);

my $reload_event = Event->signal(
	signal	=> 'HUP',
	desc	=> 'reload on SIGHUP',
	prio	=> 6,
	cb	=> sub {
		print STDERR "Reconfiguration requested\n";
		postconfig(preconfig());
	},
);
my $quit_event = Event->signal(
	signal	=> 'INT',
	cb	=> sub { 
		print STDERR "Quitting...\n";
		Event::unloop_all();
	},
);

Event::loop();

unlink($pidfile);

exit(0);

sub error
{
	my $e = shift;
	if ($do && $do eq 'stop') {
		warn $e;
	} else {
		die $e;
	}
}
		
sub usage
{
	print <<END;
Usage: $0 [ -c file ] [ -f ] { start | stop | reload | restart | help | version  | check }
 -c		specify config file (defaults to /etc/rinetd.pl.conf)
 -f		run in the foreground (don't detach)
 start		Starts a new rinetd.pl if there isn't one running already
 stop		Stops a running rinetd.pl
 reload		Causes a running rinetd.pl to reload it's config file.  Starts
		a new one if none is running.
 restart	Stops a running rinetd.pl if one is running.  Starts a new one
		irregardless.
 check		Check the configuration file and report the daemon state
END
	exit;
}

sub preconfig
{
	open(CONFIG, $configfile) 
		or error "open $configfile: $!\n";
	my %new;
	$filters{global} = {};
	my $last = 'global';
	while (<CONFIG>) {
		next if /^#/;
		next if /^$/;
		if (/^\s*(allow|deny)\s+(\S+)\s*(?:#.*)?$/) {
			my ($action, $text) = ($1, $2);
			$text =~ s/^(\d+(?:\.\d+)*)(?:\.\*)+$/$1/;
			my $block = new2 Net::Netmask ($text);
			if ($block) {
				$block->{line} = $.;
				$block->{action} = $action;
				$block->storeNetblock($filters{$last});
			} else {
				error "parse error $configfile, line $.: $Net::Netmask::error\n";
			}
			next;
		}
		/^(\S+)\s+(\w+)\s+(\S+)\s+(\w+)\s*(?:#.*)?$/
			or error "Parse error $configfile, line $.: $_";
		my ($fhost, $fport, $thost, $tport) = ($1, $2, $3, $4);
		my $new = join("\n", $fhost, $fport, $thost, $tport, $.);
		$new{$new} = undef;
		$last = $new;
		$filters{$new} = {};
	}
	close(CONFIG);
	return %new;
}

sub postconfig
{
	my (%new) = @_;
	for my $old (keys %config) {
		next if $new{$old};
		$config{$old}->shutdown 
			if $config{$old};
		delete $config{$old};
	}
	for my $new (keys %new) {
		next if $config{$new};
		$config{$new} = new RelayListener (split(/\n/, $new), $filters{$new});
	}
}

package RelayListener;

use strict;
use Net::Netmask;

sub new 
{
	my ($pkg, $fhost, $fport, $thost, $tport, $line, $filter) = @_;
	die unless $fhost;
	die unless $fport;
	die unless $thost;
	die unless $thost;

	my $self = bless {
		tohost	=> $thost,
		toport	=> $tport,
		counter	=> $counter++,
		filter	=> $filter,
		line	=> $line,
		desc	=> "Listen $fhost:$fport -> $thost:$tport",
	}, $pkg;
	$counter++;

	my $listener = IO::Event::Socket::INET->new(
		Listen		=> 20,
		Proto		=> 'tcp',
		LocalPort	=> $fport,
		LocalHost	=> $fhost,
		Description	=> "$fhost:$fport -> $thost:$tport",
		Handler		=> $self,
		Reuse		=> 1,
	);
	unless ($listener) {
		warn "Could not listen at $fhost:$fport: $!";
		return undef;
	}
	print "$self->{desc}\n" if $debug;
	$self->{listener} = $listener;
	return $self;
}

sub shutdown
{
	my ($self) = @_;
	print "SHUTDOWN $self->{desc}\n" if $debug;
	$self->{listener}->close();
	delete $self->{listener};
	$self->{shutdown} = 1;
}

sub ie_connection
{
	my ($self, $ioe) = @_;

	print "CONNECT $self->{desc}\n" if $debug;
	my $client = $ioe->accept();

	my $client_ip = $client->peerhost();

	my $filterblock = findNetblock($client_ip, $self->{filter})
		|| findNetblock($client_ip, $filters{global});

	if ($filterblock && $filterblock->{action} eq 'deny') {
		print "DENIED from $client_ip for $self->{desc}\n";
		$client->print("501 Relay denied $self->{line}\r\n");
		$client->close();
	} else {
		print "accepted from $client_ip for $self->{desc}\n";
		$client->readevents(0);
		RelayConnect->new($client, $self->{desc}, $self->{tohost}, $self->{toport});
	}
}

sub ie_input
{
	die "why?";
}


package RelayConnect;

use strict;
use POSIX qw(ETIMEDOUT);

sub new
{
	my ($pkg, $client, $desc, $tohost, $toport) = @_;

	my $self = bless {
		desc	=> $desc,
		tohost	=> $tohost,
		toport	=> $toport,
		client	=> $client,
		counter	=> $counter++,
	}, $pkg;
		
	IO::Event::Socket::INET->new(
		PeerAddr	=> $tohost,
		PeerPort	=> $toport,
		Proto		=> 'tcp',
		Handler		=> $self,
	);
	return undef;
}

sub ie_connected
{
	my ($self, $ioe) = @_;
	print "CONNECT Server$self->{counter} $self->{desc}\n" if $debug;
	if ($self->{relaylisten}{shutdown}) {
		# oh, well
		$ioe->close();
		return;
	} 
	my $relayclient = Relay->new($self, $self->{client}, 'Client', $self->{desc});
	bless $self, 'Relay';

	$self->{other}		= $relayclient;
	$self->{role}		= 'Server';
	$self->{ioe}		= $ioe;

	delete $self->{client};
	delete $self->{tohost};
	delete $self->{toport};
}

sub ie_connect_failed
{
	my ($self, $ioe, $error) = @_;
	if ($error == ETIMEDOUT) {
		print "TIMEOUT-CONNECT Server$self->{counter} $self->{desc}\n" if $debug;
		$ioe->close();
		new RelayConnect ($self->{client}, $self->{desc}, $self->{tohost}, $self->{toport});
	} else {
		print "NO-CONNECT Server$self->{counter} $self->{desc}: $error\n" if $debug;
		$self->{client}->print("500 Relay open failed\r\n");
		$self->{client}->close();
	}
}

sub ie_input
{
	die "why?";
}


package Relay;

use strict;

# also constructed by re-blessing RelayConnect objects
sub new
{
	my ($pkg, $other, $ioe, $role, $desc) = @_;
	my $self = bless {
		ioe	=> $ioe,
		other	=> $other,
		role	=> $role,
		desc	=> $desc,
		counter	=> $counter++,
	}, $pkg;
	$ioe->handler($self);
	$ioe->readevents(1);
	return $self;
}

sub close
{
	my ($self) = @_;
	$self->{ioe}->close()
		if $self->{ioe};
	my $o = delete $self->{other};
	$o->close() if $o;
}

sub ie_input
{
	my ($self, $ioe, $ibr) = @_;
	print "DATA $self->{role}$self->{counter} $self->{desc}\n" if $debug;
	if (defined $self->{other}) {
		$self->{other}{ioe}->print($$ibr) || warn "print: $!";
	} else {
		warn "other not defined";
	}
	$$ibr = '';
}

sub ie_werror
{
	my ($self, $ioe) = @_;
	print "WRITE-ERROR $self->{role}$self->{counter} $self->{desc}\n" if $debug;
	$self->close();
}

sub ie_eof
{
	my ($self, $ioe, $ibr) = @_;
	print "EOF $self->{role}$self->{counter} $self->{desc}\n" if $debug;
	$self->close();
}

sub ie_outputoverflow
{
	my ($self, $ioe, $overflowing) = @_;
	print "OVERFLOW-$overflowing $self->{role}$self->{counter} $self->{desc}\n" if $debug;
	$self->{other}{ioe}->readevents(! $overflowing)
		if $self->{other};
}

__END__

=head1 NAME

rinetd.pl - tcp redirection server

=head1 SYNOPSIS

rinetd.pl [ -c configfile ] [ -f ] { start | stop | reload | restart | check }

=head1 DESCRIPTION

Rinetd.pl forwards tcp connections from one IP address and port to
another.  rinetd.pl can forward from multiple ports simultaneously
as defined in a config file (/etc/rinetd.pl.conf).

Rinetd.pl is called "rinetd.pl" instead of simply "rinetd" so as
to not be confused with Thomas Boutell's "rinetd" program.

Exactly one of imperitive is required on the command line:

=over 9

=item start

Start a new rinetd server if there isn't one already running.

=item stop

Stop a running rinetd server

=item restart

Stop the running rinetd server (if one is running).
Start a new rinetd server.

=item reload

Reconfigure the running rinetd server.  Start a new server
if none is running.

=back

The command line options are:

=over 9

=item -c file

Specify an alternative configuration file.  Multiple rinetd.pl servers
can run simultaneously if they have different config files.

=item -f

Run in the foreground.  Normally rinetd.pl detaches itself and runs
as a deamon.  When it runs as a daemon it redirects its output through
the logger(1) program.

=back

=head1 FORWARDING RULES

The format for forwarding rules (in the config file) is:

 from-ip-address from-port to-ip-address to-port

IP addresses and ports can be numeric or named.  Use 0.0.0.0 for
listening on all IP addresses.

=head1 FILTER RULES

Allow and deny rules can control what IP addresses are allowed
to use the server.

The format of rules is:

 allow|deny netblock

Filters that follow a forwarding rule apply to that 
forwarding rule only.

Filters that preceed any forwarding rules apply to 
all forwarding rules if no per-forwarding rule filter
matches.

Filters are not ordered: the most specific filter (smallest
network block) that matches is the one that is used.

Filter rules may be indented for clarity.  Filter rules must be
numeric -- hostnames are not allowed.

=head1 EXAMPLE CONFIG

 # We have to start with the global access 
 # control list.
 # The order of the rules does not matter.

 deny	any 		# '0.0.0.0/0' and 'default' work too
 deny	216.240.32.1 
 allow	216.240.32/24

 0.0.0.0 8282 idiom.com 23

 allow	216.240.47/24
 deny	216.240.47.38
 deny	216.240.32.4

 0.0.0.0 daytime idiom.com daytime # idiom's clock is better
 
=head1 LICENSE

Copyright (C) 2005 David Muir Sharnoff <muir@idiom.com>.
This module may be used/copied/etc on the same terms as Perl 
itself.
